native/core/src/execution/datafusion/shuffle_writer.rs

```diff
@@ -954,10 +986,65 @@ impl ShuffleRepartitioner {
         });
         Ok(used)
     }
+
+    /// Appends rows of specified indices from columns into active array builders in the specified partition.
+    async fn append_rows_to_partition(
+        &mut self,
+        columns: &[ArrayRef],
+        indices: &[usize],
+        partition_id: usize,
+    ) -> Result<isize> {
+        let mut mem_diff = 0;
+
+        let output = &mut self.buffered_partitions[partition_id];
+
+        let time_metric = self.metrics.baseline.elapsed_compute();
+
+        // If the range of indices is not big enough, just appending the rows into
+        // active array builders instead of directly adding them as a record batch.
+        let mut start_index: usize = 0;
+        let mut output_ret = output.append_rows(columns, indices, start_index, time_metric);
+
+        loop {
+            match output_ret {
+                AppendRowStatus::MemDiff(l) => {
+                    mem_diff += l?;
+                    break;
+                }
+                AppendRowStatus::StartIndex(new_start) => {
+                    // Cannot allocate enough memory for the array builders in the partition,
+                    // spill partitions and retry.
+                    self.spill().await?;
+                    self.reservation.free();
```

Comment:
Forgot to free memory reservation in previous commit.

@andygrove Could you try run benchmarks again? Thanks.

Reply:
I no longer see the memory error, but there seems to be a significant performance regression. TPC_H q2 used to take 12 seconds and is now taking many minutes. I do not see spill happening in Spark UI. I am going to add some debug logging to try and understand what is happening.

Reply:
In native metrics, I do see excessive spilling:

```
 spill_count=8, spilled_bytes=19441254400, data_size=877436
 ```

Reply:
Okay. I will take a look it further.

Reply:
I guess that is because we silently use some memory but never report them into the reservation, like the memory usage on array builders, now we count for them. So under same memory settings, it is more likely you hit the bar of memory pool. Have you try to increase the Comet memory like `spark.comet.memoryOverhead`?

Reply:
I tried setting the overhead:
 
 ```
     --conf spark.executor.instances=1 \
    --conf spark.executor.memory=16G \
    --conf spark.executor.cores=8 \
    --conf spark.cores.max=8 \
    --conf spark.memory.offHeap.enabled=true \
    --conf spark.memory.offHeap.size=20g \
    --conf spark.comet.memoryOverhead=16g \
```

This did not help with performance:

```
Query 2 took 352.05119466781616 seconds
```

Reply:
With latest commit, I ran TPCH sf=100 locally and didn't see regression now. Can you verify it?

Reply:
Looks good! 

```
Query 2 took 12.09787917137146 seconds
```